checkout: show summary by default and introduce --show-changes flag

skshetry opened the pull request:
* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.
https://github.com/iterative/dvc.org/pull/1054

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Resolves #2979.

**TODO**
- [x] Add few tests.
- [x] `CheckoutError`: go with the consensus. 
- [x] Make changes to flags if required. (make default?)
- [ ] Handle `--relink` and `modified` case.
- [ ] Show `--relink` successfully message. (Separate PR)
- [ ] Show how many files were downloaded/fetched. (Separate PR)
- [ ] Remove warnings on checkout failure
- [ ] Fix pull/fetch outputs
- [ ] Autocompletion scripts and help messages

skshetry commented on dvc/remote/base.py:
```diff
@@ -966,12 +978,14 @@ def checkout(
                 )
-            return failed
+            if failed:
+                raise CheckoutError([failed])
```
A lot of tests will fail because of this and I'm really not sure if this is a good thing or bad. But, for now, it really makes the control flow easier.

skshetry replied:
Tests seem to have passed. :see_no_evil:

Suor commented on dvc/stage.py:
```diff
@@ -982,18 +982,24 @@ def checkout(
+                elif is_modified is False:
+                    checkouts["added"].append(out.path_info)
+            except CheckoutError as exc:
+                checkouts["failed"].append(*exc.target_infos)
```
Should this be `.extend()`?

Suor commented on dvc/state.py:
```diff
@@ -453,20 +453,22 @@ def remove_unused_links(self, used):
+
+    def remove_links(self, unused):
+        for path in unused:
+            remove(os.path.join(self.root_dir, path))
```
There was no join in original code, so this might be an error. BTW, paths are rarely rel to dvc root, they are either absolute or relative to cwd.

skshetry replied:
I splitted the function as-is. Original code used `path` which was absolute path. But, we might not need it.

Suor commented on tests/func/test_state.py:
```diff
@@ -69,16 +70,33 @@ def test_get_state_record_for_inode(get_inode_mock, tmp_dir, dvc):
+    tmp_dir.dvc_gen({"foo": "foo_content", "bar": "bar_content"})
+
+    with dvc.state:
+        links = [os.path.join(dvc.root_dir, link) for link in ["foo", "bar"]]
```
Why using absolute paths?

skshetry replied:
`get_unused_links` already expect it to be a abspath. So, I'd not like to change it here (at least, not in this PR).

efiop commented on dvc/command/checkout.py:
```diff
@@ -26,6 +85,18 @@ def add_parser(subparsers, parent_parser):
+        action="store_true",
+        default=False,
+        help="Show summary of changes",
+    )
```
Not sure if we need these as options, seems like we could show both by-default.

skshetry replied:
Good point! We could make it default, but, I'll wait for product feedback.

shcheklein replied:
Would love to see them by default (summary at least)! 

may be --show-json to being able to script this, but it's another PR probably

shcheklein replied:
If summary is default, then we can name option -l (--long) or something like this to show all of the output

skshetry replied:
Like we discussed below around `git checkout`, it is never silent. It'll display every changes if there's any unless silenced by `-q`.

Okay, the api already gives us a dict of checkouts. So, it should be easy to implement. But, let's keep it in another PR.

skshetry replied:
This is now the default (if there's no changes, there's no output though).

skshetry commented:
@efiop @Suor, the plan was to just show stats/summary. More verbose checkout changes did not seem too hard with existing stats/summary implementation so I quickly prototyped on it with `--show-changes`. I've asked @dmpetrov for feedback.

skshetry commented on dvc/remote/base.py:
```diff
@@ -899,27 +906,35 @@ def _checkout_dir(
                 progress_callback(str(entry_info))
 
-        self._remove_redundant_files(path_info, dir_info, force)
+        _modified = _modified or self._remove_redundant_files(
```
Nice catch! Thanks.

skshetry commented on dvc/stage.py:
```diff
@@ -982,18 +982,24 @@ def checkout(
+                )
+                if is_modified:
+                    checkouts["modified"].append(out.path_info)
+                elif is_modified is False:
```
For now, *added* is *False*, *modified* is *True*, and *no change* is *None*. :see_no_evil:

skshetry commented on dvc/repo/checkout.py:
```diff
@@ -59,13 +63,17 @@ def _checkout(
+                relink=relink,
+                filter_info=filter_info,
             )
+            failed.extend(result["failed"])
```
Missed it! Thanks.

pared commented on dvc/remote/base.py:
```diff
@@ -899,27 +906,35 @@ def _checkout_dir(
 
+        if _added or _modified:
+            return not _added and _modified
+
```
Shouldn't we return something when `_added=False` and `_modified=False`?

skshetry replied:
I understand it's confusing, and I'll probably refactor it to return some kind of enum status. I'll leave it as it is for now, till I know more requirements (or, if this suffice). Thanks for the review!

shcheklein commented on dvc/command/checkout.py:
```diff
@@ -1,18 +1,91 @@
+logger = logging.getLogger(__name__)
+
+
+def _humanize_wordseries(words=None):
```
move it to the utils? there should be some libraries that do different "humanization" stuff ... not sure if we want to utilize them.

skshetry replied:
I'd like to keep it near to where it's used for now. If we later need it elsewhere, we can do so at that time.

shcheklein replied:
👍 up to you, absolutely minor thing. 

Let me share my intuition behind this, though :)

- **Readability** - any boilerplate code that makes file longer and absolutely not relevant/generic makes it harder to see the "bigger picture" (tools like CodeClimate complain about file being too long, about too many methods, etc for a good reason);
- **Reusability** - the knowledge about this generic "text" util is only in your head. Low chances that someone will be able to find it when it's needed again. Means some probability of a duplicate later.
- **Testing** - you can unit test this simple things, would be strange to write a unit test for a private method though.

Suor replied:
It's not like humanizing, it's joining and not word series but some items in a list, so the name might be better, e.g. `human_join()` or smth.

Suor replied:
On keeping it here - I am with @skshetry here, move it later if that is used elsewhere. Spreading code is the big source of errors. Also impairs readability.

shcheklein commented on dvc/command/checkout.py:
```diff
@@ -1,18 +1,91 @@
+
+def _log_changes(stats, with_failure=False):
+    colors = [
+        ("added", colorama.Fore.GREEN),
```
does it makes to utilize LS_COLORS as well? may be check how does git do this with diffs?

JIoJIaJIu replied:
I've checked git, and it works based on `diff.color`/`color.diff` config option
[diff.c](https://github.com/git/git/blob/bfdd66e72fffd18235757bedbf355fd4176d6019/diff.c#L57)
[diff.h](https://github.com/git/git/blob/b02fd2accad4d48078671adf38fe5b5976d77304/diff.h#L389)
[diff.color.](https://github.com/git/git/blob/bfdd66e72fffd18235757bedbf355fd4176d6019/diff.c#L435)
```
$ git config --local color.diff.new yellow
$ git log -p
```
https://stackoverflow.com/a/26941235/484582
<hr/>

but gnu diffutils with [this commit](http://git.savannah.gnu.org/cgit/diffutils.git/commit/?id=04f6d57177fb6c9e7bbcf716c521bc323936a9a5) started to use theme with the same notation as ls does
But it isn't propagated from the env variable
```
defaults to @samp{rs=0:hd=1:ad=32:de=31:ln=36} for red deleted lines,
green added lines, cyan line numbers, bold header
```
```
$ echo 1 > file1
$ echo 2 > file2
$ diff file1 file2 --color --palette="ad=01;35:de=01;36"
...
```

skshetry replied:
The current output is very similar to `dvc diff`.

> does it makes to utilize LS_COLORS as well?

you mean, colorizing pathnames?

Another possible output format to use is of `git checkout`:
```sh
$ git checkout branch7
M	.gitignore
A	data2/data
Switched to branch 'branch7'
```

shcheklein replied:
I meant more of a reusing some colorization/output style of Git (diff first came to my mind with red for removed, green for added). Idea to mimic `git checkout` is also good.

Instead of hardcoding color codes utilize some environment (in a way Gut did his reserach)

skshetry replied:
We could color it after #3351 is merged. It'd be better to go with however `ls` works (and, not have too many color formats).

`git checkout` style is easily greppable, whereas `dvc diff` style output is not. 
```diff
$ dvc checkout
+ M data
! D foo
+ A models
- F something
# ignore the colors and symbols
```

We could make the pathnames follow `LS_COLORS`, whereas keep *{M,A,D,F}* as {yellow, green, orange, and red}* (or, something even better). What do you think, @shcheklein?

skshetry replied:
But, I assume `dvc diff` can be easily understood.

shcheklein replied:
I like it! Both idea to use git checkout format and LS_COLORS.

skshetry replied:
Let's discuss on `LS_COLORS` separately. I have made the output similar to `git checkout`.

JIoJIaJIu commented on dvc/remote/base.py:
```diff
@@ -899,27 +906,36 @@ def _checkout_dir(
 
+        if added or modified:
+            return not added and modified
+
```
This section is not really clear for me and it can return `None` and `bool`
Maybe just 
```python
return modified and not added
```

skshetry replied:
@JIoJIaJIu, thanks for the review. This is something that I will fix it later on (but, before, I'd like to find what's better in terms of product POV).

If you have any suggestions regarding how we could do this better, please do tell. :slightly_smiling_face:

JIoJIaJIu commented on dvc/stage.py:
```diff
@@ -982,18 +982,24 @@ def checkout(
-                failed_checkouts.append(failed)
-        return failed_checkouts
+            try:
+                is_modified = out.checkout(
```
What do you think to apply one naming, either `is_modified` or `modified everywhere?

skshetry replied:
Thanks for the suggestion. I have fixed it.

skshetry commented on dvc/command/checkout.py:
```diff
@@ -1,18 +1,77 @@
+        if self.args.show_changes:
+            _log_changes(stats, exc is not None)
+        if self.args.summary:
+            _log_summary(stats, exc is not None)
```
We do checkout other files even if some have failed, so, I see *failed* as one state of a checkout too. So, if we make `checkout` show stats by default, it'd be helpful to provide information what we already have.

skshetry replied:
Summary about fail is no longer thrown. But, other details are still shown even if there's a failure.

skshetry commented on dvc/repo/checkout.py:
```diff
@@ -8,14 +8,14 @@
         if out.scheme == "local"
     ]
-    repo.state.remove_unused_links(used)
+    return repo.state.get_unused_links(used)
```
TODO:
- [ ] `get_unused_links` returns directory name without `/` at the end

Suor replied:
Because it is not for presentation. I don't like where it's headed all internal code should not care about presentation, that will complicate things a lot.

skshetry replied:
Yes, I'm not suggesting to change internal functions at all. Just the `_get_unused_links` function inside the `repo/checkout.py`. I'd love to handle it in `command`, but as we need to return dict (as we might need to support `--show-json` later), I feel this is the place to do. 

But, I agree with you. I am not liking this at all, sprinkling logic all over the place to return changes. If you have any better suggestion, please do tell.

skshetry replied:
@Suor, we could return stats without stringifying filepaths (i.e. as PathInfo), and only add the `os.sep` on the deleted files beforehand (on `Repo._checkout()` itself). This way, presentation logic would be on the `cli` side. However, this will divide the logic anyway.

JIoJIaJIu commented on dvc/stage.py:
```diff
@@ -982,22 +983,30 @@ def checkout(
             except CheckoutError as exc:
-                checkouts["failed"].extend(exc.target_infos)
+                checkouts["failed"].extend(
+                    [target for target in _fspath_dir(exc.target_infos)]
```
is it misprint?
```python
[_fspath_dir(target) for target in exc.target_infos]
```

skshetry replied:
Nice catch! Thanks. I should write more tests. :)

JIoJIaJIu commented on dvc/stage.py:
```diff
@@ -982,22 +983,30 @@ def checkout(
+        def _fspath_dir(path):
+            """Return string with "/" at the end if it's a dir."""
+            p = os.path.join(path, "") if os.path.isdir(path) else fspath(path)
+            return relpath(p, self.repo.root_dir)
```
What do you think about it? (if I got the function purpose right)
```python
rel = relpath(p, self.repo.root_dir)
return "{}{}".format(rel, os.path.sep) if os.path.isdir(path) else rel
```

shcheklein commented:
@skshetry would be super helpful if you show some examples (just here in the comments) - how would outputs looks like in different settings, and summary of changes that were introduced.

shcheklein commented:
Will be writing questions here:

1. If checkout fails it does not show any summary of changes, even though it deletes files, may be even adds? We should first show changes, then fail.

shcheklein commented:
@skshetry please see comments above.

skshetry commented:
> 1. If checkout fails it does not show any summary of changes, even though it deletes files, may be even adds? We should first show changes, then fail.

That should already happen (infact it's tested). Does that not work for you or, you meant something like:
```
1 added, 2 deleted and 3 failed
```

Why I did not do this is because, I find the error message fairly good:
```
$ dvc checkout
WARNING: Cache '8d1069328fcefdcc8a0969744b3ab7ab' not found. File 'foo' won't be created.
1 added and 1 deleted
ERROR: unexpected error - Checkout failed for following targets:
 foo
Did you forget to fetch?
```

and,
```
$ dvc checkout --show-changes
WARNING: Cache '8d1069328fcefdcc8a0969744b3ab7ab' not found. File 'foo' won't be created.
A	data/
D	bar
ERROR: unexpected error - Checkout failed for following targets:
 foo
Did you forget to fetch?
```

> 2. I don't see any reason to introduce an option - let's make it a default behavior. It looks like it is exactly what git does, --show-changes looks great! If we need to, we can create an option to show a summary instead.

I was worried that it'll be very noisy during checkout. So, I just went with showing just the summary by default. What do you think?

> 3. I'm getting:
> 
> ```
> Nothing to do
> ...
> ```

Looks like I forgot to handle this in a different PR. I'll just remove this message. Checkout will be silent if nothing changes.

> 
> 1. Do we show a summary of changes inside directory or just `M  dir/path`? Would be really nice to see at least a summary but it can be a separate PR.

Let's do that in a separate PR. But, like above, this might get very noisy. Should we introduce a flag for these, or only when a stage file for a directory is checked out?

> 2. It affects `dvc pull`:
> 
> ```
> (.env) [ivan@ivan ~/Projects/example-get-started]$ dvc pull
> Nothing to do
> Everything is up to date.
> ```
> 
> 1. Since `checkout` is the last step of the `dvc pull` we should show the same summary (+ pull specific summary that can be a separate PR)

Working on it. Introduce `--show-changes` flag here as well?

> 2. `--relink` is shown as modified - it okay, but can be better. Modified can be confusing here.

Thanks for mentioning this. I have fixed this.

> 3. test carefully with `-R`, --with-deps`, etc - with `-R` I was getting some confusing results.

Yeah, will check. Thanks for reminding.
I'll add a few screenshots here as well after I fixed your suggestions.

cc @shcheklein

shcheklein commented:
@skshetry thanks!

> That should already happen (infact it's tested). Does that not work for you or, you meant something like:

I will check again, but most likely you are right. Most likely I just missed this message. Both warnings are way more colorful and distract attention - you just don't see the summary.

 > Does that not work for you or, you meant something like

Number of failed would be good, but mostly I meant that I didn't see the message at all and thought that we missed it.

Do we need that warning? Also, if we switch the mode to the `--show-changed` one, it can eliminate the problem (partially). 

>  Introduce --show-changes flag here as well?

as I mentioned, I don't see the need for `--show-changes`. I would make it a default mode. I think we are overcomplicating this a bit. `--show-changes` works perfectly and behave like Git. The same for `pull`.

> Thanks for mentioning this. I have fixed this.

yes, we should show `--relink` is a different way, otherwise we have M for both - relinked and modified files - not sure this is a good idea.

skshetry commented:
> Looks good @skshetry!
> 2 things:
> Main question:
> 
> 1. `fetch` - don't we want to show a summary of how much files have been downloaded?
>

I'll prefer to keep those on different PR. Let's keep the scope of this to checkout for now (pull only shows info about checkout). I'll create a separate issue later on.

> Side comment:
> 2. Regarding @shcheklein comments on dir contents:
> I think we should create issue for that. We are currently treating all outs as single entities during checkout, so check out for 1M files directory and single file out will show `2 changed`. I think the next step could be to introduce something like `SyncResult` object, that will take care of gathering information on what has been done during `fetch/checkout/pull` operations. That way we would not have to return tuples of booleans from sync methods.

I really did not like introducing a *global* object. Instead of booleans, we could use some kind of `Status` enum object that is collected upwards and then summarized/processed.

skshetry commented:
> Do we need that warning? Also, if we switch the mode to the `--show-changed` one, it can eliminate the problem (partially).

What do you think of checkout being noisy? What's your view on this, @shcheklein?

> yes, we should show `--relink` is a different way, otherwise we have M for both - relinked and modified files - not sure this is a good idea.

Okay, I won't extend the scope of this PR. For now, `--relink` is no-change (so, no outputs unless anything other also got modified on the workspace). 

We can just have `relinked successfully` message on the end if `--relink` is used. I think, this should be okay (unless `--show-changes` is specified, where no information is show for relinked).

shcheklein commented:
@skshetry I don't see any good reason to show that warning (along with the `Checkout failed for following targets:` message). Also, I don't see any good reason to have a separate `--show-changes` - I would make it default - it imitates git behavior and provides really good and clean output.

> For now, --relink is no-change 

agreed, this is good! the most important is not show all of the as modified.

> We can just have relinked successfully message on the end

agreed again! should be enough.

skshetry commented:
> I don't see any good reason to show that warning (along with the `Checkout failed for following targets:` message).

So, you meant something like:
```sh
$ dvc checkout --show-changes
A	data/
D	bar
F     foo
$ echo $?
1
```

> Also, I don't see any good reason to have a separate `--show-changes` - I would make it default - > it imitates git behavior and provides really good and clean output.

So, how'd this work? Should we need summary to show by default if we also show changes by default? This will make outputs ungreppable, so, we made them work separately.

shcheklein commented:
> So, you meant something like:

I'm not sure about F to be honest. If we consider this an error - let's show that last error message. But I would remove WARNING that happen before output.

> This will make outputs ungreppable, so, we made them work separately.

@skshetry probably, no summary then - do we actually need it?

(btw, my comparison with git checkout is wrong, but it does not show summary as well, so it's probably better to not compare it at all)

skshetry commented:
I'd love to hear other's opinion on this as well. cc @iterative/engineering 

I think, showing every change on the checkout is too verbose (maybe I'm optimizing for large changes by default) and would only like to show summary:
 ```sh
dvc checkout
A	data/
D	bar
```

vs

```sh
dvc checkout
1 added and 1 deleted
```

pared commented:
@skshetry @shcheklein I would say that writing changes are not too verbose until we are staying with the current approach to summary (which is calculating them on `outs`). If we will start to aggregate, for example, the whole contents of the directory, that might be too much.

My reasoning for that is that the number of outs is limited indirectly by user, who "marks" outs with `dvc` commands.

skshetry commented:
Another thing in favor of showing complete changes by default is if user has hooks for `git checkout`. This will have pretty similar output formats.

casperdcl commented:
I prefer `+-` over `AD`, and `?` or `E` over `F`

pared commented on dvc/repo/checkout.py:
```diff
@@ -1,21 +1,31 @@
+    if not os.path.exists(path):
+        return str(path)
+
+    path = relpath(fspath(path), root)
```
Looking at tests for `3.5` path should be string here, which is not always the case.
If we have `path` in arguments, I think we should always put path here and not PathInfo object. Now it is mixed.

skshetry commented on tests/func/test_checkout.py:
```diff
@@ -513,5 +520,230 @@ def test_checkout_relink_protected(tmp_dir, dvc, link):
+        Bucket=remote.path_info.bucket, Key=file_path.path, Body="foo\nfoo"
+    )
+    stats = dvc.checkout(force=True)
+    assert stats == {**empty_checkout, "modified": [str(file_path)]}
```
I only added this tests after your review, @pared.

pared replied:
Thanks for pointing, makes review easier :)

skshetry commented:
@shcheklein, I had a discussion with Dmitry. The objective of the PR was to:
> To my mind, we should NOT make checkout silent.

I feel this PR does more than that already. So, I'd like to move forward with the PR as-is for now and keep the discussions ongoing. What do you say? We can reach a conclusion later on and could be handled on separate PRs.

shcheklein commented:
@skshetry travis is failing?

> I feel this PR does more than that already. 

I agree, that's why I was surprised to see two additional optional initially :)

If we keep `--show-changes` (up to you, but I would still reconsider this - remove or make it default - I just don't see much of a reason in this option, and it was not part of the requirement), please don't forget to update:

- autocomplete scripts - bash and zsh
- docs
- ask @jorgeorpinel to review the help message

Also, let me do the final pass - I'll try to run and check the PR today.

shcheklein commented on dvc/exceptions.py:
```diff
@@ -256,7 +256,9 @@ def __init__(self, amount):
+        self.stats = stats
         targets = [str(t) for t in target_infos]
         m = (
             "Checkout failed for following targets:\n {}\nDid you "
```
sorry, really minor request (feel free to nor change this), came out as I was reviewing:

```
ERROR: unexpected error - Checkout failed for following targets:
 data/prepared
data/features
data/data.xml
```

there is an extra space that is not needed before `data/prepared`. I think the change is `\n {}\n` -> `\n{}\n`

shcheklein commented:
@skshetry going back to the requirements in the #2979 (you mention in the description) , I think it does not resolve them for the failed case:

```
WARNING: Cache '6836f797f3924fb46fcfd6b9f6aa6416.dir' not found. File 'data/prepared' won't be created.
WARNING: Cache 'a304afb96060aad90176268345e10355' not found. File 'data/data.xml' won't be created.
WARNING: Cache '42c7025fc0edeb174069280d17add2d4.dir' not found. File 'data/features' won't be created.
1 added
ERROR: unexpected error - Checkout failed for following targets:
 data/prepared
data/data.xml
data/features
Did you forget to fetch?

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

we still have excessive WARNINGS . We can create a separate PR for this or keep the ticket open. I think a summary is enough:

```
1 added, 3 failed

ERROR: unexpected error - Checkout failed for following targets:
 data/prepared
data/data.xml
data/features
Did you forget to fetch?
```

Again, happy to just keep that ticket open.

shcheklein commented:
It still affects `dvc pull` in some unexpected (?) ways:

```
(.env) [ivan@ivan ~/Projects/example-get-started]$ dvc pull
2 added
Everything is up to date.
```

shcheklein commented:
(Also, going through the example get started I have more and more reasons for the `--show-changes` option by default. One more reason - we mix directories and files when we count added, modified, etc. E.g. when it prints `1 added` but actually it's a directory with two preprocessed files it make it confusing if `dvc checkout model.dvc` prints the same. Just a thought to consider.)

skshetry commented:
> it's a directory with two preprocessed files it make it confusing if `dvc checkout model.dvc` prints the same. Just a thought to consider.)

I'm also more inclined to make this the default, even more so with making output formats of `git` and `dvc` checkouts similar (helpful with `post-checkout`). So, to not make anymore changes, I'll swap the defaults, and introduce `--summary` flag.

Later, we could decide if we really need `--summary`, and then either remove or even rename those as `--shortstat`/ `--short`/`--stat`/`--numstat`, etc.

> It still affects `dvc pull` in some unexpected (?) ways:

Looks like on second example, `pull` did fetch something and did checkout (checkout looks to be working fine), but `pull` did not report anything (same in-line with previous behavior that it didn't print anything if it fetched something). I'll fix `fetch` and `pull` behavior on separate PRs.

> we still have excessive WARNINGS . We can create a separate PR for this or keep the ticket open. I think a summary is enough:

It'd be better to just reduce the level of loggings here to `DEBUG`. Again, I'll handle it in a separate PRs (added to TODO).

Thanks Ivan for checking out!

Suor commented:
@skshetry I don't think you can simply make those warnings debug, this will affect more things than just checkout, e.g. `Stage.run()` does not check for those checkouts to succeed.

Suor commented on dvc/external_repo.py:
```diff
@@ -106,7 +106,11 @@ def _pull_cached(self, out, path_info, dest):
+            try:
+                out.checkout(filter_info=src)
+            except CheckoutError:
+                failed = True
```
Looks like there is an error here introduced earlier, one shouldn't move/remove if there is already an error. Simply raise `FileNotFoundError` here instead of using `failed` flag.

Suor commented on dvc/remote/base.py:
```diff
@@ -851,20 +852,25 @@ def safe_remove(self, path_info, force=False):
         if progress_callback:
             progress_callback(str(path_info))
 
+        # relink is not modified
```
You don't really know that relink is not modified, so this is untrue. I see that you are going to handle relink later, but still there should not be any lying comments in code.

Suor commented on dvc/repo/checkout.py:
```diff
@@ -52,20 +72,23 @@ def _checkout(
-    if failed:
-        raise CheckoutError(failed)
+            for data in ["failed", "added", "modified"]:
+                stats[data] += [
```
```suggestion
                stats[data].extend(
```
No need to create an intermediate list nor rewrite a source.

Suor commented on dvc/repo/checkout.py:
```diff
@@ -34,9 +44,19 @@ def _checkout(
+        unused = _get_unused_links(self)
+
+    stats["deleted"] = [_fspath_dir(u, self.root_dir) for u in unused]
+    self.state.remove_links(unused)
```
Why is this not under if? Looks fragile this way, we shouldn't call `remove_links()` at all if a particular target is specified.

skshetry replied:
It's empty if there's no argument.

skshetry replied:
Also, I need to check if something is a dir before deleting.

Suor replied:
There is also an argument, it might be an empty sequence, but it's there.

Suor replied:
Checkout in dvc:
- checks out particular thing(s) if that things (target) is specified
- if no target is specified it checks out everything and removes anything previously checked out and having no references now.

I.e. if there is target we don't try to change the rest of the workspace, so `.remove_links()` should not be called. The logic of your code is different.

skshetry replied:
If targets are specified, `unused` will be empty and hence, nothing will get removed. If your concern is regarding calling `.remove_links()`, I don't think there's any reason to be defensive.

skshetry commented:
@shcheklein, I have changed the defaults. There's a `--summary` flag for showing summary instead of detaliled changelog. Other stuffs will be fixed in a separate PR.

shcheklein commented:
sorry, wanted to approve initially, missed the button )

jorgeorpinel commented on dvc/command/checkout.py:
```diff
@@ -26,6 +94,12 @@ def add_parser(subparsers, parent_parser):
+        "--summary",
+        action="store_true",
+        default=False,
+        help="Show summary of the changes",
```
```suggestion
        help="Show summary of the changes.",
```

skshetry commented:
@shcheklein, @efiop, I have created an issue https://github.com/iterative/dvc/issues/3483 for handling autocompletion scripts.
